pkg/generic/thriftidl_provider.go

```diff
@@ -84,29 +128,58 @@ const defaultMainIDLPath = "main.thrift"
 func NewThriftContentProvider(main string, includes map[string]string) (*ThriftContentProvider, error) {
 	p := &ThriftContentProvider{
 		svcs: make(chan *descriptor.ServiceDescriptor, 1), // unblock with buffered channel
+		opts: &ProviderOption{DynamicGoEnabled: false},
 	}
-	tree, err := ParseContent(defaultMainIDLPath, main, includes, false)
+	svc, err := newServiceDescriptorFromContent(defaultMainIDLPath, main, includes, false)
 	if err != nil {
 		return nil, err
 	}
-	svc, err := thrift.Parse(tree, thrift.DefaultParseMode())
+
+	p.svcs <- svc
+	return p, nil
+}
+
+// NewThriftContentProviderWithDynamicGo builder
+func NewThriftContentProviderWithDynamicGo(main string, includes map[string]string) (*ThriftContentProvider, error) {
+	p := &ThriftContentProvider{
+		svcs: make(chan *descriptor.ServiceDescriptor, 1), // unblock with buffered channel
+		opts: &ProviderOption{DynamicGoEnabled: true},
+	}
+	svc, err := newServiceDescriptorFromContent(defaultMainIDLPath, main, includes, false)
 	if err != nil {
 		return nil, err
 	}
+
+	if err = newDynamicgoDscFromContent(svc, defaultMainIDLPath, main, includes, false); err != nil {
+		// fall back to the original way (without dynamicgo)
+		p.opts.DynamicGoEnabled = false
+		klog.CtxWarnf(context.Background(), "KITEX: failed to get dynamicgo service descriptor, fall back to the original way, error=%s", err)
+		return p, nil
+	}
+
 	p.svcs <- svc
 	return p, nil
 }
 
 // UpdateIDL ...
 func (p *ThriftContentProvider) UpdateIDL(main string, includes map[string]string) error {
+	var svc *descriptor.ServiceDescriptor
 	tree, err := ParseContent(defaultMainIDLPath, main, includes, false)
 	if err != nil {
 		return err
 	}
-	svc, err := thrift.Parse(tree, thrift.DefaultParseMode())
+	svc, err = thrift.Parse(tree, thrift.DefaultParseMode())
 	if err != nil {
 		return err
 	}
+
+	if p.opts.DynamicGoEnabled {
```

Comment:
this block looks identical to L153 block?

Reply:
done